cookbook/models.py

```diff
@@ -37,16 +37,73 @@ def get_model_name(model):
 
 
 class TreeManager(MP_NodeManager):
+    # model.Manager get_or_create() is not compatible with MP_Tree
     def get_or_create(self, **kwargs):
-        # model.Manager get_or_create() is not compatible with MP_Tree
         kwargs['name'] = kwargs['name'].strip()
-        q = self.filter(name__iexact=kwargs['name'], space=kwargs['space'])
-        if len(q) != 0:
-            return q[0], False
-        else:
+        try:
+            return self.get(name__iexact=kwargs['name'], space=kwargs['space']), False
+        except self.model.DoesNotExist:
             with scopes_disabled():
-                node = self.model.add_root(**kwargs)
-            return node, True
+                return self.model.add_root(**kwargs), True
+
+
+class TreeModel(MP_Node):
+    _full_name_separator = ' > '
+
+    def __str__(self):
+        if self.icon:
+            return f"{self.icon} {self.name}"
+        else:
+            return f"{self.name}"
+
+    @property
+    def parent(self):
+        parent = self.get_parent()
+        if parent:
+            return self.get_parent().id
+        return None
+
+    @property
+    def full_name(self):
+        """
+        Returns a string representation of a tree node and it's ancestors,
+        e.g. 'Cuisine > Asian > Chinese > Catonese'.
+        """
+        names = [node.name for node in self.get_ancestors_and_self()]
+        return self._full_name_separator.join(names)
+
+    def get_ancestors_and_self(self):
+        """
+        Gets ancestors and includes itself. Use treebeard's get_ancestors
+        if you don't want to include the node itself. It's a separate
+        function as it's commonly used in templates.
+        """
+        if self.is_root():
+            return [self]
+        return list(self.get_ancestors()) + [self]
+
+    def get_descendants_and_self(self):
+        """
+        Gets descendants and includes itself. Use treebeard's get_descendants
+        if you don't want to include the node itself. It's a separate
+        function as it's commonly used in templates.
+        """
+        return self.get_tree(self)
+
+    def has_children(self):
+        return self.get_num_children() > 0
+
+    def get_num_children(self):
+        return self.get_children().count()
+
+    # use self.objects.get_or_create() instead
+    @classmethod
+    def add_root(self, **kwargs):
+        with scopes_disabled():
```

Comment:
same as above

Reply:
the scopes_disabled might be redundant in one of these spots, left over when I was reorganizing the code.